File: lib/ch4/4.3.js
// Given a sorted (increasing order) array, write an algorithm to
// create a binary tree with minimal height.

class Node {
  constructor(data) {
    this.data = data;
    this.left = null;
    this.right = null;
  }
}

function createBinaryTree(arr, start = 0, end = arr.length) {
  if (start > end) return null;

  const midx = Math.floor(start + end / 2);
  const root = new Node(arr[midx]);
  root.left = createBinaryTree(arr, start, midx - 1);
  root.right = createBinaryTree(arr, midx + 1, end);

  return root;
}

File: java/ch4/4.3.java
class BTreeNode {
    BTreeNode left;
    BTreeNode right;
    int data;
    public BTreeNode(int data) {
        this.data = data;
        this.left = left;
        this.right = right;
    }
}

BTreeNode sortedArrToBTree(int[] nums, int start, int end) {
    if (start > end) return null;
    int mid = (start + end) / 2;
    int d = nums[mid];
    BTreeNode root = new BTreeNode(d);
    root.left = sortedArrToBTree(nums, start, mid - 1);
    root.right = sortedArrToBTree(nums, mid + 1, end);
    return root;
}
